File: spa/src/lib/relay.ts
// HTTP client for the buzz relay bridge (POST /query, POST /events),
// mirroring buzz-cli: ws(s):// normalized to http(s)://, per-request NIP-98.

import type { Signer, SignedEvent } from "./nostr";

export function normalizeRelayUrl(url: string): string {
  let base = url.trim().replace(/\/+$/, "");
  if (base.startsWith("ws://")) base = `http://${base.slice(5)}`;
  else if (base.startsWith("wss://")) base = `https://${base.slice(6)}`;
  else if (!/^https?:\/\//.test(base)) base = `https://${base}`;
  return base;
}

export class Relay {
  private readonly base: string;

  constructor(url: string, private readonly signer: Signer) {
    this.base = normalizeRelayUrl(url);
  }

  private async post(path: string, payload: unknown): Promise<Response> {
    const url = `${this.base}${path}`;
    const body = JSON.stringify(payload);
    return fetch(url, {
      method: "POST",
      headers: {
        Authorization: await this.signer.nip98Header("POST", url, body),
        "Content-Type": "application/json",
      },
      body,
    });
  }

  async query(filters: Record<string, unknown>[]): Promise<SignedEvent[]> {
    const resp = await this.post("/query", filters);
    if (!resp.ok) {
      throw new Error(`query failed (${resp.status}): ${(await resp.text()).slice(0, 300)}`);
    }
    return (await resp.json()) as SignedEvent[];
  }

  async submit(event: SignedEvent): Promise<unknown> {
    const resp = await this.post("/events", event);
    if (!resp.ok) {
      throw new Error(`publish failed (${resp.status}): ${(await resp.text()).slice(0, 300)}`);
    }
    return resp.json().catch(() => ({}));
  }
}

File: app/relay.py
"""HTTP client for the buzz relay bridge endpoints (POST /query, POST /events).

Mirrors crates/buzz-cli/src/client.rs: ws(s):// URLs are normalized to
http(s)://, every request carries a freshly signed NIP-98 Authorization
header, filters are POSTed as a JSON array, writes POST one signed event.
"""

from __future__ import annotations

import json

import httpx

from .nostr import Signer


class RelayError(Exception):
    pass


def normalize_relay_url(url: str) -> str:
    url = url.strip().rstrip("/")
    if url.startswith("ws://"):
        url = "http://" + url[len("ws://") :]
    elif url.startswith("wss://"):
        url = "https://" + url[len("wss://") :]
    elif not url.startswith(("http://", "https://")):
        url = "https://" + url
    return url


class Relay:
    def __init__(self, url: str, signer: Signer, timeout: float = 15.0):
        self.base = normalize_relay_url(url)
        self.signer = signer
        self._http = httpx.Client(timeout=timeout)

    def _post(self, path: str, payload) -> httpx.Response:
        url = f"{self.base}{path}"
        body = json.dumps(payload, separators=(",", ":"), ensure_ascii=False).encode()
        headers = {
            "Authorization": self.signer.nip98_header("POST", url, body),
            "Content-Type": "application/json",
        }
        try:
            return self._http.post(url, content=body, headers=headers)
        except httpx.HTTPError as exc:
            raise RelayError(f"relay unreachable: {exc}") from exc

    def query(self, filters: list[dict]) -> list[dict]:
        resp = self._post("/query", filters)
        if resp.status_code != 200:
            raise RelayError(f"query failed ({resp.status_code}): {resp.text[:300]}")
        data = resp.json()
        if not isinstance(data, list):
            raise RelayError(f"unexpected query response: {str(data)[:300]}")
        return data

    def submit(self, event: dict) -> dict:
        resp = self._post("/events", event)
        if resp.status_code not in (200, 201, 202):
            raise RelayError(f"publish failed ({resp.status_code}): {resp.text[:300]}")
        try:
            return resp.json()
        except ValueError:
            return {}
